### ISSUE TITLE
jtidy does not handle <textarea> correctly

### ISSUE THREAD

**Issue by desb42** [gnosygnu/xowa#250]
Taking the page '/home/wiki/Special:XowaCfg?grp=xowa.html.page' via the xowa web server
![options_html_page](https://user-images.githubusercontent.com/861631/47147461-2d4a4600-d2c6-11e8-9fbb-bf4f36bc18e4.jpg)

If you look at the source code generated (around the 'Content code format' area)
![source_code](https://user-images.githubusercontent.com/861631/47147834-fcb6dc00-d2c6-11e8-8fc7-1adc06ea070e.jpg)

towards the bottom of the page is a <textarea> section
Note that the content of the <textarea> actually comes after the closing </textarea>

The actual page does not seem to be affected - but this is not so.
An ajax call in the background requests this page's data and the page is overwritten with the correct html. 

The problem is that as the html is being constructed on the xowa side the html is passed through a jtidy step. It is this step that is corrupting the html
The ajax call does not get passed through a jtidy step which is why that is correct.

A solution I would suggest is to not do a jtidy step if it is a 'Special' page (namespace -1)
This would be independent of the option to turn jtidy on or off
(Of course correcting jtidy behaviour is also another possibility)

**Comment by gnosygnu** [comment]
Hi,

Thanks for the detail, and a nice catch on your side. This looks like a bug on the http_server side. The XOWA GUI works fine, but the HTTP_SERVER is applying jtidy before it serves the page

I think your workaround is reasonable. I have some vacation time coming up and will take a look at it then.

Thanks again

----

Link for XOWA GUI: home/wiki/Special:XowaCfg?grp=xowa.gui.menus

**Comment by desb42** [comment]
Thanks for the response (by the way a great piece of code)

I do not think this is an http_server issue. As far as I can tell both the gui and http_server hit the same code

As I said originally, this issue is obscured by the fact of an ajax call immediately the page is downloaded. (downloaded into the gui or a 'real' web browser)

As this ajax call takes place every time. Why go to the trouble of generating the html (by a server side mustache module written in java), when the ajax call overwrites with a javascript mustache function?

**Comment by gnosygnu** [comment]
Yeah, my mistake in recollection. I wrote this code a few years ago, and based my response on my memories of its intent.

> As this ajax call takes place every time. Why go to the trouble of generating the html (by a server side mustache module written in java), when the ajax call overwrites with a javascript mustache function?

I think I did this so that there would be some HTML available in the oddball-case that the AJAX call fails. Ultimately, the AJAX/mustache calls needs to work in order for the Special:XowaCfg to function properly

I think your suggestion could work (skipping HTML Tidy if Special:XowaCfg). That said, it looks like the page still functions correctly as long as the AJAX call works. (I thought it wasn't working in http_server mode) It looks like the main issus is that the HTML markup in "View Source" is incorrect?

**Comment by desb42** [comment]
The issue I am trying to point out is that jtidy messes up the html when the html includes <textarea>
eg
html == <textarea>some text</textarea>
when passed thru jtidy
html == <textarea></textarea>some text
So not calling jtidy avoids the issue for Special pages

**Comment by gnosygnu** [comment]
Yeah, I guess I was saying that the messed-up html has no real user-impact. The Options page shows the value correctly inside the &lt;textarea> tags (due to the AJAX mustache call) and the messed-up html only appears when viewing the HTML without the AJAX (like ```view-source:http://localhost:8080/home/wiki/Options```)

That said, I agree that this behavior is sloppy and unintuitive. Skipping jtidy for the Special pages works for now. I think a longer term fix may be to escape content inside &lt;textarea> tags. See: https://sourceforge.net/p/jtidy/bugs/131/

Specificially, note that this problem only occurs b/c the textarea has an XML node: ```<pre>~{page_text}</pre>```. If there was no XML node (for example, ```~{page_text}``` then the text would be properly rendered inside the &lt;textarea> tags

Anyway, I'll incorporate your commit in the jtidy-fix branch when I get back (https://github.com/desb42/xowa/commit/9d529523356389733624e19bb85518caea8aa7fc). Thanks

**Comment by gnosygnu** [comment]
I pushed a commit now to escape textarea content. I tested this locally, and it correctly handles the ```<pre>~{page_text}</pre>``` above

I'd like to start with this specific fix for now. If the messed-up textarea shows up elsewhere, I'll try to push a fix similar to the above. Keep in mind that it looks like textarea should only contain character data: https://stackoverflow.com/questions/2580247/can-i-embed-html-formatting-inside-of-a-textarea-tag . As such, I'd rather see how far we get with the "fix them as we see them" approach, rather than "skip jtidy if in Special namespace" one. 

Thanks
